src/main/java/graphql/annotations/processor/graphQLProcessors/GraphQLInputProcessor.java

```diff
@@ -52,27 +49,27 @@ public GraphQLInputProcessor() {
      * @return a {@link GraphQLInputType} that represents that object class
      */
 
-    public GraphQLInputType getInputType(Class<?> object, ProcessingElementsContainer container) {
+    public GraphQLInputType getInputTypeOrRef(Class<?> object, ProcessingElementsContainer container) {
         boolean considerAsInput = true;
         if (Enum.class.isAssignableFrom(object)) {
             considerAsInput = false;
         }
 
-        String typeName = graphQLObjectInfoRetriever.getTypeName(object);
+        String typeName = DEFAULT_INPUT_PREFIX +graphQLObjectInfoRetriever.getTypeName(object);
 
-        if (considerAsInput)
-            typeName = DEFAULT_INPUT_PREFIX + typeName;
+        if (container.getProcessing().contains(typeName)){
+            return new GraphQLTypeReference(typeName);
+        }
 
         Map<String, GraphQLType> typeRegistry = container.getTypeRegistry();
 
         if (typeRegistry.containsKey(typeName)) {
             return (GraphQLInputType) container.getTypeRegistry().get(typeName);
         } else {
 
-            graphql.schema.GraphQLType graphQLType = graphQLOutputObjectRetriever.getGraphQLType(object, container, considerAsInput);
-            GraphQLInputType inputObject = graphQLInputObjectRetriever.getInputObject(graphQLType, DEFAULT_INPUT_PREFIX, typeRegistry);
-            typeRegistry.put(inputObject.getName(), inputObject);
-            return inputObject;
+            GraphQLInputType graphQLType = (GraphQLInputType) graphQLTypeRetriever.getGraphQLType(object, container, considerAsInput);
```

Comment:
Does the cast work on enums? because they're not considered as input

Reply:
EnumBuilder returns a GraphQLEnumType which implements both GraphQLOutputType and GraphQLInputType, so the cast works